adafruit_ads7830.py

```diff
@@ -26,7 +25,127 @@
+        0x07,  # DIFF_CH7_CH6
+    ]
+
+    def __init__(self, i2c: I2C, address: int = _I2C_ADDR) -> None:
+        self.i2c_device = I2CDevice(i2c, address)
+
+    def _single_channel(self, channel: int, pd_mode: int = 3) -> int:
+        """ADC value in single-ended mode
+        Scales the 8-bit ADC value to a 16-bit value
+
+        :param int channel: Channel (0-7)
+        :param int pd_mode: Power-down mode
+        :return: Scaled ADC value or raise an exception if read failed
+        :rtype: int
+        """
+        if channel > 7:
+            raise ValueError("Invalid channel: must be 0-7")
+
+        command_byte = self._CHANNEL_SELECTION[channel]
+        command_byte <<= 4
+        command_byte |= self._POWER_DOWN_MODES[pd_mode] << 2
+
+        with self.i2c_device as i2c:
+            try:
+                # Buffer to store the read ADC value
+                adc_value = bytearray(1)
+                i2c.write_then_readinto(bytearray([command_byte]), adc_value)
+                # Scale the 8-bit value to 16-bit
+                return adc_value[0] << 8
+            except Exception as error:
+                raise RuntimeError(f"Failed to read value: {error}") from error
+
+    def _differential_channel(self, channel: int, pd_mode: int = 3) -> int:
+        """ADC value in differential mode
+        Scales the 8-bit ADC value to a 16-bit value
+
+        :param int channel: Positive channel for differential reading (0-7)
+        :param int pd_mode: Power-down mode
+        :return: Scaled ADC value or raise an exception if read failed
+        :rtype: int
+        """
+        if channel > 7:
+            raise ValueError("Invalid channel: must be 0-7")
+
+        command_byte = self._DIFF_CHANNEL_SELECTION[channel // 2]
+        command_byte <<= 4
+        command_byte |= self._POWER_DOWN_MODES[pd_mode] << 2
+
+        with self.i2c_device as i2c:
+            try:
+                # Buffer to store the read ADC value
+                adc_value = bytearray(1)
+                i2c.write_then_readinto(bytearray([command_byte]), adc_value)
+                # Scale the 8-bit value to 16-bit
+                return adc_value[0] << 8
+            except Exception as error:
+                raise RuntimeError(f"Failed to read value: {error}") from error
+
+    @property
+    def value(self) -> typing.List[int]:
+        """Single-ended mode ADC values for all channels
+
+        :rtype: List[int]"""
+        values = []
+        for i in range(8):
+            single_value = self._single_channel(i)
+            values.append(single_value)
+        return values
+
+    @property
+    def differential_value(self) -> typing.List[int]:
+        """Differential ADC values for all channel pairs
+
+        :rtype: List[int]"""
+        values = []
+        for i in range(0, 8, 2):  # Iterate over channel pairs
+            diff_value = self._differential_channel(i)
+            values.append(diff_value)
+        return values
```

Comment:
I wouldn't have these read every value, every time because that'll be slow and do a bunch of work that the user may not need. Instead, take a look at this library: https://github.com/adafruit/Adafruit_CircuitPython_MCP3xxx/tree/main/adafruit_mcp3xxx

It provides a way to get the value directly and also has an AnalogIn compatibility class too.

Reply:
awesome, thanks! that was really helpful to reference